### PR TITLE
fix(relationships): can now prevent relationships using event handler

### PR DESCRIPTION
**Pull request by beck24**
###### Todo
- [x] Add tests
- [x] Add a note to documentation

### CODE REVIEW

**Comment by beck24** [comment]
It seems we have no test coverage for any relationships stuff - this includes some initial tests for relationships as well as cases that would have caught this issue.

File: engine/tests/ElggRelationshipTest.php [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+	public function testAddRelationship() {
+		// test adding a relationship
+		add_entity_relationship($this->entity1->guid, 'test_relationship', $this->entity2->guid);
+		
+		$this->assertTrue(check_entity_relationship($this->entity1->guid, 'test_relationship', $this->entity2->guid));
+	}
+	
+	public function testRemoveRelationship() {
+		// test adding a relationship
+		add_entity_relationship($this->entity1->guid, 'test_relationship', $this->entity2->guid);
```
Comment (mrclay) [inline engine/tests/ElggRelationshipTest.php]: Should we test return value?

File: engine/tests/ElggRelationshipTest.php [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+		remove_entity_relationship($this->entity1->guid, 'test_relationship', $this->entity2->guid);
+		
+		$this->assertFalse(check_entity_relationship($this->entity1->guid, 'test_relationship', $this->entity2->guid));
+	}
+	
+	public function testPreventAddRelationship() {
+		// test event handler - should prevent the addition of a relationship
+		elgg_register_event_handler('create', 'relationship', array($this, 'preventRelationship'));
+		
+		add_entity_relationship($this->entity1->guid, 'test_relationship', $this->entity2->guid);
```
Comment (mrclay) [inline engine/tests/ElggRelationshipTest.php]: Same here.

File: engine/tests/ElggRelationshipTest.php [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+		$this->assertTrue(check_entity_relationship($this->entity1->guid, 'test_relationship', $this->entity2->guid));
+		
+		remove_entity_relationship($this->entity1->guid, 'test_relationship', $this->entity2->guid);
+		
+		$this->assertFalse(check_entity_relationship($this->entity1->guid, 'test_relationship', $this->entity2->guid));
+	}
+	
+	public function testPreventAddRelationship() {
+		// test event handler - should prevent the addition of a relationship
+		elgg_register_event_handler('create', 'relationship', array($this, 'preventRelationship'));
```
Comment (mrclay) [inline engine/tests/ElggRelationshipTest.php]: Use `Elgg\Values::getFalse` for these handlers.

**Comment by mrclay** [comment]
If there are docs for the create relationship event we may want to mention there that returning `false` was broken until (whatever version)

**Comment by mrclay** [comment]
I like that the PHP docs mention stuff like that anyway.

**Comment by beck24** [comment]
This should be good now, can I get another once-over from someone?

**Comment by hypeJunction** [comment]
In general LGTM. I think I would prefer class assertions (assertIsA) in stead of true (assertTrue) for `check_entity_relationship()` calls.
It would perhaps be wise to test that relationships are deleted when entity is deleted, especially when there is a handler preventing relationship from being deleted.

**Comment by beck24** [comment]
Great suggestions, thanks

File: docs/guides/events-list.rst [hunk 4]
```diff
@@ -105,11 +105,14 @@ User events
 Relationship events
 ===================
 
-**create, <relationship>**
+**create, relationship**
```
Comment (mrclay) [inline docs/guides/events-list.rst]: 1.12 still fires two events. We should probably do this:
```
**create, relationship**
**create, <relationship_name>**
```

Comment (mrclay) [inline docs/guides/events-list.rst]: Same thing for the delete event(s)

Comment (beck24) [inline docs/guides/events-list.rst]: We can, I assumed we wouldn't want deprecated calls in the official docs though.

Comment (mrclay) [inline docs/guides/events-list.rst]: You're right.

**Comment by beck24** [comment]
Updated as per IK's suggestions, plus added tests for the following entity methods

    addRelationship
    removeRelationship
    deleteRelationships

**Comment by mrclay** [comment]
Great work on the tests

**Pull request merged**